linkerd/app/outbound/src/http/logical/policy.rs

```diff
@@ -0,0 +1,396 @@
+}
+
+// === impl Routes ===
+
+impl Routes {
+    pub fn addr(&self) -> &Addr {
+        match self {
+            Routes::Http(RouterRoutes { ref addr, .. })
+            | Routes::Grpc(RouterRoutes { ref addr, .. }) => addr,
+        }
+    }
+}
+
+// === impl Params ===
+
+impl<T> Params<T>
+where
+    // Parent target type.
+    T: Clone + Debug + Eq + Hash + Send + Sync + 'static,
+{
+    /// HTTP or gRPC policy routing.
+    pub(super) fn layer<N, S>() -> impl svc::Layer<
+        N,
+        Service = svc::ArcNewService<
+            Self,
+            impl svc::Service<
+                    http::Request<http::BoxBody>,
+                    Response = http::Response<http::BoxBody>,
+                    Error = Error,
+                    Future = impl Send,
+                > + Clone,
+        >,
+    > + Clone
+    where
+        // Inner stack.
+        N: svc::NewService<Concrete<T>, Service = S>,
+        N: Clone + Send + Sync + 'static,
+        S: svc::Service<
+            http::Request<http::BoxBody>,
+            Response = http::Response<http::BoxBody>,
+            Error = Error,
+        >,
+        S: Clone + Send + Sync + 'static,
+        S::Future: Send,
+    {
+        svc::layer::mk(|inner: N| {
+            let http = svc::stack(inner.clone()).push(RouterParams::layer());
+            let grpc = svc::stack(inner).push(RouterParams::layer());
+
+            http.push_switch(
+                |pp: Params<T>| {
+                    Ok::<_, Infallible>(match pp {
+                        Self::Http(http) => svc::Either::A(http),
+                        Self::Grpc(grpc) => svc::Either::B(grpc),
+                    })
+                },
+                grpc.into_inner(),
+            )
+            .push(svc::ArcNewService::layer())
+            .into_inner()
+        })
+    }
+}
+
+impl<T> From<(Routes, T)> for Params<T>
+where
+    T: Eq + Hash + Clone + Debug,
+{
+    fn from((pr, parent): (Routes, T)) -> Self {
+        match pr {
+            Routes::Http(http) => Params::Http(RouterParams::from((http, parent))),
+            Routes::Grpc(grpc) => Params::Grpc(RouterParams::from((grpc, parent))),
+        }
+    }
+}
+
+impl<T> svc::Param<super::LogicalAddr> for Params<T>
+where
+    T: Eq + Hash + Clone + Debug,
+{
+    fn param(&self) -> super::LogicalAddr {
+        match self {
+            Params::Http(p) => p.param(),
+            Params::Grpc(p) => p.param(),
+        }
+    }
+}
+
+// === impl RouterParams ===
+
+impl<T, M, F> RouterParams<T, M, F>
+where
+    // Parent target type.
+    T: Clone + Debug + Eq + Hash + Send + Sync + 'static,
+    // Request matcher.
+    M: Clone + Send + Sync + 'static,
+    // Request filter.
+    F: Eq + Hash,
+    F: Clone + Send + Sync + 'static,
+{
+    // Wraps a `NewService`--instantiated once per logical target--that caches a set
+    // of concrete services so that, as the watch provides new `Params`, we can
+    // reuse inner services.
```

Comment:
tiny nit, not a blocker: if this is a doc comment, it will be visible if we build RustDoc with `--document-private-items` --- so it may be slightly preferable to use a real doc comment even though the function is public.
```suggestion
    /// Wraps a `NewService`--instantiated once per logical target--that caches a set
    /// of concrete services so that, as the watch provides new `Params`, we can
    /// reuse inner services.
```